skrub/_table_vectorizer.py

```diff
@@ -495,147 +449,95 @@ def _clone_transformers(self):
-                                warnings.warn(
-                                    "Warning raised by pandas when converting column"
-                                    f" '{col}' to datetime: "
-                                    + str(warning.message),
-                                    UserWarning,
-                                    stacklevel=2,
-                                )
-                    except (ValueError, TypeError):
-                        pass
-            # Cast pandas dtypes to numpy dtypes
-            # for earlier versions of sklearn. FIXME: which ?
-            if issubclass(X[col].dtype.__class__, ExtensionDtype):
-                try:
-                    X[col] = X[col].astype(X[col].dtype.type, errors="ignore")
-                except (TypeError, ValueError):
-                    pass
-            self.types_[col_idx] = X[col].dtype
-        return X
+            X[col] = _replace_missing_indicators(X[col])
+
+            # Some numerical dtypes like Int64 or Float64 only support
+            # pd.NA, so they must be converted to np.float64 before imputing
+            # with np.nan.
+            if is_numeric_dtype(X[col]) and X[col].isna().any():
+                X[col] = X[col].astype(np.float64)
+
+            # Cast pandas dtypes to numpy dtypes for earlier versions of sklearn.
+            # Categorical dtypes don't need to be casted.
+            # Note that 'is_category_dtype' is deprecated.
+            if is_extension_array_dtype(X[col]) and not isinstance(
+                X[col].dtype, CategoricalDtype
+            ):
+                dtype = X[col].dtype.type
+                X[col] = X[col].astype(dtype)
+
+                # When converting string to object, <NA> values becomes '<NA>'
+                # so we need to replace false missing values once more.
+                X[col] = _replace_missing_indicators(X[col])
+
+            # For object dtype columns, convert to string to avoid mixed types.
+            if is_object_dtype(X[col]):
+                mask = X[col].notnull()
+                X.loc[mask, col] = X.loc[mask, col].astype(str)
+
+        if reset:
+            X = to_datetime(X)
+            X = _to_numeric(X)
+            self.inferred_column_types_ = X.dtypes.to_dict()
 
-    def _apply_cast(self, X: pd.DataFrame) -> pd.DataFrame:
-        """Takes a dataframe, and applies the best data types learnt during fitting.
+        else:
+            # Update categorical dtype.
+            error_msg = (
+                "This %(name)s instance is not fitted yet. "
+                "Please run auto_cast(X, reset=True) first."
+            )
+            check_is_fitted(self, ["inferred_column_types_"], msg=error_msg)
+            category_columns = X.select_dtypes("category").columns
+            for col in category_columns:
+                dtype = self.inferred_column_types_[col]
+                dtype = _union_category(X[col], dtype)
```

Comment:
I'm not sure we should update the type to add categories in transform but let's deal with it in #848

Reply:
I'm not 100% sure of the rational of doing it, we should discuss it